src/details/ArborX_DetailsTreeConstruction.hpp

```diff
@@ -202,10 +203,17 @@ class GenerateHierarchy
     //   concatenation.
     // In this case, if the Morton indices are the same, we want to compare is.
     // We also want the result in this situation to always be less than any
-    // Morton comparison. Thus, we add INT_MIN to it.
-    // We also avoid if/else statement by doing a "x + !x*<blah>" trick.
+    // Morton comparison. Thus, we add LLONG_MIN to it.
     auto x = _sorted_morton_codes(i) ^ _sorted_morton_codes(i + 1);
-    return x + (!x) * (INT_MIN + (i ^ (i + 1)));
+    if (x != 0)
+    {
+      // When using 63 bits for Morton codes, the LLONG_MAX is actually a valid
+      // code. As we want the return statement above to return a value always
+      // greater than anything here, we downshift by 1.
+      return x - 1;
+    }
+
+    return LLONG_MIN + (i ^ (i + 1));
```

Comment:
I am not following why this cannot just be
```C++
return x + (!x) * (LLONG_MIN + (i ^ (i + 1)));

Reply:
Consider a tree with two points, `(0,0,0)` and `(1,1,1)`. The corresponding Morton indices are 0 and `LLONG_MAX`. During construction, the thread for the first leaf (index 0), computed `delta_left = LLONG_MAX` (due to the early exit check in the `delta()`, and `delta_right = LLONG_MAX` (due to `0 ^ LLONG_MAX = LLONG_MAX`). Then, `is_left_child` is set to 0 as `!(delta_right < delta_left)`. This ends up be further in `left_child = -1`, which is completely invalid.

Reply:
Admittedly, this is not the prettiest workaround. We could use 60-bit Morton indices instead, which would avoid that.

Reply:
I am missing something.  I thought for 32-bit Morton code we were expanding by 10 and for 64-bit by 21 which I would expect means we only use 63 bits.

Reply:
Andrey pointed out to me integral type was signed